File: Dart/longest-valid-parentheses.dart
class Solution {
  int longestValidParentheses(String s) {
    int result = 0;
    for (int i = 0; i < s.length; i++) {
      if (s[i] == ')') continue;
      int max = 0;
      int stack = 0;
      for (int j = i; j < s.length; j++) {
        if ('(' == s[j]) {
          stack++;
          max++;
          continue;
        }
        if (0 == stack) {
          break;
        }
        stack--;
        max++;
        if (0 == stack && result < max) {
          i = j + 1;
          result = max;
        }
      }
    }
    return result;
  }
}

void main() {
  Solution solution = Solution();
  print(solution.longestValidParentheses("(())(")); // Output: 4
  print(solution.longestValidParentheses(")()())")); // Output: 4
  print(solution.longestValidParentheses("")); // Output: 0
}

File: C/longest-valid-parentheses.c
#include <string.h>

int longestValidParentheses(char *s) {
    int result = 0;
    int length = strlen(s);

    for (int i = 0; i < length; i++) {
        int max = 0;
        int stack = 0;

        for (int j = i; j < length; j++) {
            if (s[j] == '(') {
                stack++;
                max++;
                continue;
            }
            
            if (stack == 0) {
                break;
            }

            stack--;
            max++;

            if (stack == 0 && result < max) {
                result = max;
            }
        }

        if (stack > 0) {
            continue;
        }
    }

    return result;
}
